#29 добавить кэши в сервис заказов

Vasyap99 opened the pull request.

Commit:
added cache-0

Commit:
#1 исправил аннотации кеширования #2 добавил тест для кеширования

Commit:
исправил кеширование update

Commit:
исправления стиля кода

RinatBeybutov commented on orderService/src/test/java/com/maxima/orderService/config/TestContainersConfigForProducts.java:
```diff
@@ -0,0 +1,22 @@
+ * Класс конфигурации контейнеров
+ */
+@Testcontainers
+public class TestContainersConfigForProducts {
```
лучше заменить на TestContainersConfig

Vasyap99 replied:
Если унаследовать оба класса тестов от одного TestContainersConfig, то при сборке на стадии тестирования вылетает ошибка. Один тест проходит, другой вылетает, а если от двух отдельных классов, то нормально собирается

RinatBeybutov replied:
https://stackoverflow.com/questions/70100554/how-to-make-a-common-testcontainer-for-multiple-jupiter-tests

RinatBeybutov commented on orderService/src/test/java/com/maxima/orderService/controller/ProductApiIntegrationalTests.java:
```diff
@@ -151,4 +153,45 @@ void testDeleteSuccess() {
+            ProductViewDto.class);
+    log.info("<<testCache:9>>");
+
+    var deleteResponse = restTemplate.exchange(url + "/" + product.getUuid(),
```
в библиотеке мокито есть функциональность для проверки количества раз вызова какого-либо метода. Давай проверим, что при помещении в кэш метод репозитория не вызывается

Vasyap99 replied:
я попробовал, у меня пока не полуается.
беру productService, внеряемый Spring-ом,  вставляю mock класса ProductRepository через reflection, потом выполняю методы create() и find() сервиса, но методы моего мок-объекта почему-то не выполяются. выполняются методы самого репозитория. не нашел пока способ, как правильно подставить мок репозитория, чтобы подсчитать количество вызовов

`
  try {
    System.out.println(">>>>>>1");
    ProductController productController = new ProductController(productService);

    Field field = null;
    Field mapperField = null;
    ProductRepository pr1 = null;

    Class<?> cls = productService.getClass();

    for(Field f : FieldUtils.getAllFieldsList(cls)) {
      System.out.println( ">>>:"+f.getType().getName()+"::"+f.getName()+":" );
      if(f.getName().equals("repository")){
        System.out.println( ":"+f.getType().getName());
        field=f;
        f.setAccessible(true);
        pr1 = (ProductRepository) f.get(productService);
      }
      if(f.getName().equals("categoryRepository")){
        System.out.println( ":"+f.getType().getName());
        f.setAccessible(true);
        f.set(productService,categoryRepository);
      }
      if(f.getName().equals("mapper")){
        System.out.println( ":"+f.getType().getName());
        mapperField=f;
        f.setAccessible(true);
        mapperField.set(productService,mapper);
      }
    }


    ///Field field = productService.getClass().getField("repository");
    field.setAccessible(true);
    ProductRepository pr0 = (ProductRepository) field.get(productService);
    System.out.println(">>>>>>2:"+pr0+pr1);

    ProductRepository pr = Mockito.mock(ProductRepository.class);

    field.set(productService,pr);

    System.out.println(":-:"+field.get(productService));

    Mockito.doAnswer(invocation -> {
      System.out.println(">>>>MOCK!");
      return productRepository.save(invocation.getArgument(0));
    }).when(pr).save(any());

    Mockito.doAnswer(invocation -> {
      System.out.println(">>>>MOCK!");
      return productRepository.getByUuid(invocation.getArgument(0));
    }).when(pr).getByUuid(any());

    System.out.println(">>>>>>2.8");

    ProductService ps1=Mockito.mock(ProductService.class);

    final Field field1=field;
    Mockito.doAnswer(invocation -> {
      System.out.println(">>>>MOCK1!");
      field1.set(productService,pr);
      return productService.create(invocation.getArgument(0));
    }).when(ps1).create(any());

    Mockito.doAnswer(invocation -> {
      System.out.println(">>>>MOCK1!");
      System.out.println(":-:::"+field1.get(productService));
      field1.set(productService,pr);
      return productService.find(invocation.getArgument(0));
    }).when(ps1).find(any());


    System.out.println(">>>>>>3");
    // Имитация вызовов
    ProductViewDto e=ps1.create(productCreateDto());
    System.out.println(">>>>>>3.1");
    ps1.find(e.getUuid());
    System.out.println(">>>>>>3.2");
    ps1.find(e.getUuid());
    System.out.println(">>>>>>4");

  // Верификация: метод getByUuid был вызван 1 раз
    Mockito.verify(pr, Mockito.times(1)).getByUuid(e.getUuid());
  // Верификация: метод ни разу не был вызван
  //Mockito.verify(productService, Mockito.never()).getOrderCount();
    System.out.println(">>>>>>5");
  }catch (Exception ee){
    System.out.println(">>>>>>");
    ee.printStackTrace();
  }`

RinatBeybutov replied:
1. рефлексию выкинь, она здесь не нужна. 
2. тестирование кэша вынеси в отдельный класс
3. Используй мок репозитория
@MockBean
  private ProductRepository repository;

4. Дальше нужно мокнуть получение сущности, те метод repository.getByUuid. 
5. Ну и в конце ты проверяшь сколько раз вызывался метод поиска сущности у репозитория. На этом тест заканчивается. И еще нужно внедрить CacheManager и проверить там наличие закэшированного значения
6. Можно создать второй тест, в котором проверяешь удаление записи из кэша при обновлении
7. Можно еще кэш

RinatBeybutov commented on orderService/src/main/java/com/maxima/orderService/service/ProductServiceImpl.java:
```diff
@@ -36,7 +42,9 @@ public ProductViewDto create(ProductCreateDto dto) {
 
   @Override
   @Transactional(readOnly = true)
+  //@Cacheable(cacheNames = "products")
```
удали лишний комментарий

RinatBeybutov replied:
см https://habr.com/ru/articles/465667/
И еще добавь cachePut для создания и обновления.

RinatBeybutov commented on orderService/src/main/java/com/maxima/orderService/service/ProductServiceImpl.java:
```diff
@@ -26,7 +30,9 @@ public class ProductServiceImpl implements ProductService {
   @Transactional
+  @CacheEvict(cacheNames = "products")
   public ProductViewDto create(ProductCreateDto dto) {
+    log.info("<<creating product:>>");
```
и еще момент: давай не будем добавлять логи. В целом логи - это хорошо, конкретно тут не сильно сложная логика и по ошибке мы итак поймем, что происходит

RinatBeybutov committed:
#29
- Добавлен тест для кэша
- фикс генерации контроллера по опен апи
- фикс запуска нескольких тестов с тестконтейнером

RinatBeybutov committed:
Merge branch 'develop' into task/29-add-caches-in-order-service

# Conflicts:
#	metricsService/src/main/java/com/maxima/metricsService/api/MetricsApiDelegateImpl.java
#	metricsService/src/test/java/com/maxima/metricsService/MetricsServiceApplicationTests.java
#	orderService/src/test/java/com/maxima/orderService/controller/CategoryApiIntegrationTests.java

RinatBeybutov committed:
#29
- Мелкий рефакторинг
